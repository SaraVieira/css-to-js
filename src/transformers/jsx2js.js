import { formatObject } from "../utils";

/**
 * Transforms React props to a JS object.
 * @param {string} jsx props in JSX format as a single string
 * @returns {string} formatted JS object as a string
 */
export function transform(jsx) {
  // Parse JSX into a JS object
  const rules = {};
  const matchIterator = jsx.matchAll(/(\w+)=?(".*?"|{.*?}[\s}]*)?/g);
  for (const match of matchIterator) {
    let property = match[1];
    let value = match[2];

    if (property) {
      property = property.trim();
    }
    if (value) {
      value = value.trim();
    }

    if (property === "") {
      // Skip this rule
      continue;
    }

    // A property without a value is interpreted as the boolean `true`
    // (how it works in JSX)
    if (value === undefined) {
      value = "true";
    }

    // If value is an expression, remove the curly braces
    if (value.startsWith("{")) {
      value = value.slice(1, -1);
    }

    // Show undefined instead of nothing
    if (value === "") {
      value = "undefined";
    }

    rules[property] = value;
  }

  const objString = `{
    ${Object.keys(rules)
      .map(property => `${property}: ${rules[property]},`)
      .join("\n")}
  }`;

  try {
    return formatObject(objString);
  } catch (e) {
    return objString;
  }
}
