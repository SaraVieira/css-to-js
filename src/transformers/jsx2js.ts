import { formatJsObject } from "../formatters";
import { parseJsx, nodeToString } from "../parsers";

/**
 * Transforms JSX props to a JS object.
 * @param jsx props in JSX format as a single string
 */
export function transform(jsx: string): string {
  const [jsxElement, rawLines] = parseJsx(jsx);

  const propertyStrings = jsxElement.openingElement.attributes.map(
    (attribute) => {
      if (attribute.type === "JSXSpreadAttribute") {
        const argument = nodeToString(attribute.argument, rawLines);
        return `...${argument}`;
      }

      const keyString = attribute.name.name;

      let valueString: string;
      if (attribute.value) {
        if (attribute.value.type === "JSXExpressionContainer") {
          valueString = nodeToString(attribute.value.expression, rawLines);
        } else {
          valueString = nodeToString(attribute.value, rawLines);
        }
      } else {
        // Attribute without a value is shorthand for `true`
        valueString = "true";
      }

      return `${keyString}: ${valueString}`;
    }
  );

  let objString = propertyStrings.join(",\n");
  objString = `{ ${objString} }`;

  try {
    return formatJsObject(objString);
  } catch (e) {
    return objString;
  }
}
