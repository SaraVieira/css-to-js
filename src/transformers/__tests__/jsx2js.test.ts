import { transform } from "../jsx2js";

describe("jsx2js", () => {
  test("transforms a rule with a string value", () => {
    const input = `someProp="someValue"`;
    expect(transform(input)).toContain(`someProp: "someValue"`);
  });

  test("transforms a rule with a number value", () => {
    const input = `someProp={42}`;
    expect(transform(input)).toContain(`someProp: 42`);
  });

  test("transforms a rule with an object value", () => {
    const input = `someProp={{ key: "value" }}`;
    expect(transform(input)).toContain(`someProp: { key: "value" }`);
  });

  test("transforms a rule with a complex object value", () => {
    const input = `someProp={{ x: { y: z } }}`;
    expect(transform(input)).toContain(`someProp: { x: { y: z } }`);
  });

  test("transforms a rule with an arbitrary expression value", () => {
    const input = `someProp={someExpression}`;
    expect(transform(input)).toContain(`someProp: someExpression`);
  });

  test("transforms a rule without a value to `true`", () => {
    const input = `someProp`;
    expect(transform(input)).toContain(`someProp: true`);
  });

  test("transforms a spread element property", () => {
    const input = `{...someObject}`;
    expect(transform(input)).toMatchInlineSnapshot(`"{ ...someObject }"`);
  });

  test("transforms props on a single line", () => {
    expect(transform(`display="block" fontSize={16}`)).toMatchInlineSnapshot(
      `"{ display: \\"block\\", fontSize: 16 }"`
    );
  });

  test("transforms props on multiple lines", () => {
    expect(
      transform(
        `display="block"
        fontSize={16}
        margin={{ 
          xs: 4, 
          sm: 8
        }}
        padding={[2, 3]}
        {...someObject}
        fontFamily="'Inter', sans-serif" fontWeight="bold"`
      )
    ).toMatchInlineSnapshot(`
      "{
        display: \\"block\\",
        fontSize: 16,
        margin: {
          xs: 4,
          sm: 8,
        },
        padding: [2, 3],
        ...someObject,
        fontFamily: \\"'Inter', sans-serif\\",
        fontWeight: \\"bold\\",
      }"
    `);
  });
});
