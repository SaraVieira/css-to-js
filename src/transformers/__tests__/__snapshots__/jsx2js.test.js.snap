// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`jsx2js transforms props on a single line 1`] = `
"{
  display: \\"block\\",
  fontSize: 16
}"
`;

exports[`jsx2js transforms props on multiple lines 1`] = `
"{
  display: \\"block\\",
  fontSize: 16,
  margin: { xs: 4, sm: 8 },
  padding: [2, 3],
  background: \\"#1e2f5d\\",
  color: \\"#a4cff4\\",
  fontFamily: \\"'Inter', sans-serif\\",
  fontWeight: \\"bold\\"
}"
`;
